### PR TITLE
DDF-2628: Address left over comments from original SolrCloud PR

### PR DESCRIPTION
**Pull request by lessarderic**
#### What does this PR do?
Addressed unaddressed comments from https://github.com/codice/ddf/pull/1426.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@Lambeaux 
@oconnormi 
@brjeter 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[Solr](https://github.com/orgs/codice/teams/solr)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@figliold
@pklinef

#### How should this be tested? (List steps with links to updated documentation)
Use new solr.* properties in etc/system.properties to verify that all available system properties work as intended. This includes trying the embedded, HTTP, and Cloud modes as well as changing the data directory. You will need to restart DDF after changing the system.properties file. You can also set the system property in the console and restart the Solr Provider bundle.

#### Any background context you want to provide?
See https://github.com/codice/ddf/pull/142

#### What are the relevant tickets?
[DDF-2467](https://codice.atlassian.net/browse/DDF-2467)
[DDF-2628](https://codice.atlassian.net/browse/DDF-2628)

#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: distribution/ddf-common/src/main/resources/etc/system.properties [hunk 1]
```diff
@@ -23,71 +23,133 @@
...
+# Uncomment the properties based on the desired configuration:
+
+# For Http Solr Client:
 solr.client = HttpSolrClient
 solr.http.url = ${org.codice.ddf.system.protocol}${org.codice.ddf.system.hostname}:${org.codice.ddf.system.httpsPort}/solr
+solr.data.dir = ${karaf.home}/data/solr
+
+# For Cloud Solr Client:
+# solr.client = CloudSolrClient
 # solr.data.dir = ${karaf.home}/data/solr
```
Comment (figliold) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: I don't think we need solr.data.dir for the CloudSolrClient.

File: distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc [hunk 2]
```diff
@@ -15,6 +15,26 @@ Some notable features of the Solr Catalog Provider include:
 
 === Embedded Solr Catalog Provider Pros and Cons
 
+// TODO - Talk to Phil to see what he wants in this section
```
Comment (figliold) [inline distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc]: I assume the TODO will be removed.

File: distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc [hunk 3]
```diff
@@ -15,6 +15,26 @@ Some notable features of the Solr Catalog Provider include:
...
+|Scalability
+|* Does not scale. Only runs on a single server instance.
+ * Does not allow the middle tier to be scaled.
+|* Does not scale. Only runs on a single server instance.
+ * Does not allow the middle tier to be scaled.
+|* Can be scaled to as many nodes as necessary.
+
+|===
+
+// TODO - Add information about the system.properties file
```
Comment (figliold) [inline distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc]: I assume the TODO will be removed.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/ConfigurationFileProxy.java [hunk 4]
```diff
@@ -94,20 +96,26 @@ void writeSolrConfiguration(String core) {
...
      */
     public File getDataDirectory() {
         return this.dataDirectory;
     }
 
+    /**
+     * Gets the URL of a configuration file given the file and Solr core names.
+     *
+     * @param configFilename name of the configutation file to get
+     * @param core           name of the Solr core name
```
Comment (figliold) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/ConfigurationFileProxy.java]: Should this say "name of the Solr core" (without the last name)?

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java [hunk 5]
```diff
@@ -195,10 +242,10 @@ private static CloseableHttpClient getHttpClient(boolean retryRequestsOnError) {
...
 
     private static SSLContext getSslContext() {
-        if (System.getProperty("javax.net.ssl.keyStore") == null ||
-                System.getProperty("javax.net.ssl.keyStorePassword") == null ||
-                System.getProperty("javax.net.ssl.trustStore") == null ||
-                System.getProperty("javax.net.ssl.trustStorePassword") == null) {
+        if (System.getProperty("javax.net.ssl.keyStore") == null || System.getProperty(
+                "javax.net.ssl.keyStorePassword") == null || System.getProperty(
+                "javax.net.ssl.trustStore") == null || System.getProperty(
+                "javax.net.ssl.trustStorePassword") == null) {
```
Comment (brjeter) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java]: Did the formatter format this this way? Maybe it's just a preference thing, but I think this was more readable beforehand.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientFactoryImpl.java [hunk 6]
```diff
@@ -32,7 +41,11 @@
             factory = new HttpSolrClientFactory();
         }
 
-        return factory.newClient(core);
+        return newClient(factory, core);
     }
 
+    // Package-private for unit testing.
```
Comment (Lambeaux) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientFactoryImpl.java]: Was this not possible with constructor injection? Maybe we could inject a `Supplier` like last time?

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 7]
```diff
@@ -82,29 +75,64 @@ private static int parseSystemProperty(String propertyKey, int defaultValue) {
...
+     * @param collection     name of the collection to create
+     * @return {@code Future} used to retrieve the new {@link CloudSolrClient} instance
+     */
     public static Future<SolrClient> getClient(String zookeeperHosts, String collection) {
-        RetryPolicy retryPolicy = new RetryPolicy()
-                .retryWhen(null)
-                .withBackoff(10,
-                TimeUnit.MINUTES.toMillis(1),
-                TimeUnit.MILLISECONDS);
+        RetryPolicy retryPolicy = new RetryPolicy().retryWhen(null)
```
Comment (Lambeaux) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java]: Do we need the call to `.retryWhen(null)` for this to function properly?

File: distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc (hunk 2 above)
Comment (bantillo) [inline distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc]: documentation is in this PR review:  https://github.com/codice/ddf/pull/1478

File: distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc (hunk 3 above)
Comment (bantillo) [inline distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc]: will be addressed in a different PR

File: distribution/ddf-common/src/main/resources/etc/system.properties (hunk 1 above)
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: It should default to `${karaf.home}/data/solr` if not set. See `ConfigurationFileProxy` line 75.

File: distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc (hunk 3 above)
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_solr-contents/integrating-solr-contents.adoc]: Yes. I'll remove those changes from this PR.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/ConfigurationFileProxy.java (hunk 4 above)
Comment (lessarderic) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/ConfigurationFileProxy.java]: Fixed

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java (hunk 5 above)
Comment (lessarderic) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java]: Added `//` to force line breaks and improve formatting.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientFactoryImpl.java (hunk 6 above)
Comment (lessarderic) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientFactoryImpl.java]: Reworked to use a `BiFunction` instead of a template method.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java (hunk 7 above)
Comment (lessarderic) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java]: Looks like we do: https://github.com/jhalterman/failsafe#retry-policies.

File: distribution/ddf-common/src/main/resources/etc/system.properties [hunk 8]
```diff
@@ -23,71 +23,133 @@
...
 # solr.cloud.zookeeper = zookeeperhost1:9983,zookeeperhost2:9983,zookeeperhost3:9983
 
+# For Embedded Solr Client:
+# solr.client = EmbeddedSolrServer
+
```
Comment (pklinef) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: The embedded Solr client factory also uses the solr.data.dir property.

File: distribution/ddf-common/src/main/resources/etc/system.properties (hunk 1 above)
Comment (pklinef) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: Dan is right that the Solr Cloud client factory does not currently use that system property.  It does default to the value you listed but the only way to change the location used by that factory is to change `karaf.home`.  HTTP and embedded factories will override the default if `solr.data.dir` is set.  The cloud client factory does not do that currently.

I think it is fine either way.  Solr Cloud will not store data in that default location.  It just writes out the files there before uploading to Solr Cloud.  But it would not hurt to make it consistent either.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 9]
```diff
@@ -82,29 +75,64 @@ private static int parseSystemProperty(String propertyKey, int defaultValue) {
...
                 .get(() -> createSolrCloudClient(zookeeperHosts, collection));
     }
 
+    private static ScheduledExecutorService createExecutorService() throws NumberFormatException {
+        Integer threadPoolSize = parseSystemProperty("org.codice.ddf.system.threadPoolSize",
+                THREAD_POOL_DEFAULT_SIZE);
+        return Executors.newScheduledThreadPool(threadPoolSize);
+    }
+
+    private static int parseSystemProperty(String propertyKey, int defaultValue) {
```
Comment (pklinef) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java]: I think we could replace this function implementation with `NumberUtils.toInt(System.getProperty(propertyKey), defaultValue)`.

https://commons.apache.org/proper/commons-lang/javadocs/api-3.5/org/apache/commons/lang3/math/NumberUtils.html#toInt-java.lang.String-int-

File: distribution/ddf-common/src/main/resources/etc/system.properties (hunk 1 above)
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: I updated the `SolrCloudClientFactory` to use the `solr.data.dir` system property like the two other client factories. Good catch @figliold!

**Comment by lessarderic** [comment]
https://jenkins.phx.connexta.com/job/DDF-Linux-PR/24 was successful. Waiting for PR to be tested  before merging.

**Comment by lessarderic** [comment]
Manual verification successful.
